### PR TITLE
[SPARK-18188] add checksum for blocks of broadcast

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

A TorrentBroadcast is serialized and compressed first, then splitted as fixed size blocks, if any block is corrupt when fetching from remote, the decompression/deserialization will fail without knowing which block is corrupt. Also, the corrupt block is kept in block manager and reported to driver, so other tasks (in same executor or from different executor) will also fail because of it. 

This PR add checksum for each block, and check it after fetching a block from remote executor, because it's very likely that the corruption happen in network. When the corruption happen, it will throw the block away and throw an exception to fail the task, which will be retried.

Added a config for it: `spark.broadcast.checksum`, which is true by default.

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by rxin** [comment]
cc @joshrosen and @zsxwing too

File: core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala [hunk 1]
```diff
@@ -85,10 +86,21 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
...
+  /** The checksum for all the blocks. */
+  private var checksums: Array[Int] = _
+
   override protected def getValue() = {
     _value
   }
 
+  private def caclChecksum(block: ByteBuffer): Int = {
+    // block is HeapByteBuffer
+    assert(block.hasArray)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala]: Hm..., it's surprising to me that this doesn't fail any test. It means we may have some unnecessary memory copy that converts DirectByteBuffer (from Netty) to HeapByteBuffer when fetching from remote block manager.

Anyway, I think it's better to handle the non-array case like this:
```
if (block.hasArray) {
   ...
} else {
   byte[] bytes = new byte[block.remaining()]
   block.duplicate.get(bytes)
   adler.update(bytes)
}

```

File: core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala [hunk 2]
```diff
@@ -135,6 +149,10 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
         case None =>
           bm.getRemoteBytes(pieceId) match {
             case Some(b) =>
+              val sum = caclChecksum(b.chunks(0))
+              if (sum != checksums(pid)) {
+                throw new SparkException(s"corrupt remote block $pid: $sum != ${checksums(pid)}")
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala]: nit: add `broadcastId` to the error message

**Review by zsxwing — CHANGES_REQUESTED** [review]
Looks good overall. Just some nits.

@rxin do we want to add an interface for checksum in this PR?

File: core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala [hunk 3]
```diff
@@ -85,10 +86,21 @@ private[spark] class TorrentBroadcast[T: ClassTag](obj: T, id: Long)
...
   private val numBlocks: Int = writeBlocks(obj)
 
+  /** The checksum for all the blocks. */
+  private var checksums: Array[Int] = _
+
   override protected def getValue() = {
     _value
   }
 
+  private def caclChecksum(block: ByteBuffer): Int = {
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala]: `calcChecksum` ?

**Comment by davies** [comment]
Manually test this patch with a job that usually failed with corrupt streams:
```
136	26993	0	FAILED	PROCESS_LOCAL	0 / 10.1.108.161	2016/11/20 08:39:11	7 s	98 ms	0 ms	30.1 MB / 144811	29.2 MB	java.io.IOException: org.apache.spark.SparkException: corrupt remote block 9: 927759120 != 787315496 +details
java.io.IOException: org.apache.spark.SparkException: corrupt remote block 9: 927759120 != 787315496
	at org.apache.spark.util.Utils$.tryOrIOException(Utils.scala:1283)
	at org.apache.spark.broadcast.TorrentBroadcast.readBroadcastBlock(TorrentBroadcast.scala:192)
	at org.apache.spark.broadcast.TorrentBroadcast._value$lzycompute(TorrentBroadcast.scala:66)
	at org.apache.spark.broadcast.TorrentBroadcast._value(TorrentBroadcast.scala:66)
	at org.apache.spark.broadcast.TorrentBroadcast.getValue(TorrentBroadcast.scala:93)
	at org.apache.spark.broadcast.Broadcast.value(Broadcast.scala:70)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.init(Unknown Source)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8.apply(WholeStageCodegenExec.scala:367)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8.apply(WholeStageCodegenExec.scala:364)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsWithIndex$1$$anonfun$apply$25.apply(RDD.scala:820)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsWithIndex$1$$anonfun$apply$25.apply(RDD.scala:820)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:319)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:283)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:319)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:283)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:319)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:283)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:70)
	at org.apache.spark.scheduler.Task.run(Task.scala:86)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.apache.spark.SparkException: corrupt remote block 9: 927759120 != 787315496
	at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$org$apache$spark$broadcast$TorrentBroadcast$$readBlocks$1.apply$mcVI$sp(TorrentBroadcast.scala:154)
	at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$org$apache$spark$broadcast$TorrentBroadcast$$readBlocks$1.apply(TorrentBroadcast.scala:139)
	at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$org$apache$spark$broadcast$TorrentBroadcast$$readBlocks$1.apply(TorrentBroadcast.scala:139)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at org.apache.spark.broadcast.TorrentBroadcast.org$apache$spark$broadcast$TorrentBroadcast$$readBlocks(TorrentBroadcast.scala:139)
	at org.apache.spark.broadcast.TorrentBroadcast$$anonfun$readBroadcastBlock$1.apply(TorrentBroadcast.scala:204)
	at org.apache.spark.util.Utils$.tryOrIOException(Utils.scala:1276)
	... 25 more
136	27059	1	SUCCESS	PROCESS_LOCAL	13 / 10.1.109.49	2016/11/20 08:39:23	19 s	0.4 s	1 ms	30.1 MB / 144811	28.9 MB	
```

This patch did save a stage from a corrupt block of broadcast.

**Comment by zsxwing** [comment]
Can we add a flag for this feature so that people can disable it if their network is stable.

**Comment by davies** [comment]
@zsxwing Added.

**Comment by zsxwing** [comment]
LGTM pending tests

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
LGTM. Merging to master and 2.1.

**Pull request closed without merging**